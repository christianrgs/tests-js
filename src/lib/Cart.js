import find from 'lodash/find'
import remove from 'lodash/remove'

export default class Cart {
  items = []

  add(item) {
    const findItemBySkuId = cartItem => {
      return cartItem.sku.id === item.sku.id
    }

    const itemAlreadyAdded = find(this.items, findItemBySkuId)

    if (!!itemAlreadyAdded) {
      remove(this.items, findItemBySkuId)

      const itemWithQuantityUpdated = {
        ...item,
        quantity: item.quantity + itemAlreadyAdded.quantity
      }

      this.items.push(itemWithQuantityUpdated)

      return
    }

    this.items.push(item)
  }

  remove(skuId) {
    const findItemBySkuId = cartItem => {
      return cartItem.sku.id === skuId
    }

    remove(this.items, findItemBySkuId)
  }

  getTotal() {
    const totalizer = (total, item) => {
      const itemTotal = item.quantity * item.sku.price

      return total + itemTotal
    }

    const total = this.items.reduce(totalizer, 0)

    return total
  }

  getSummary() {
    const total = this.getTotal()
    const items = this.items

    return {
      items,
      total
    }
  }

  checkout() {
    const { total, items } = this.getSummary()

    this.items = []

    return {
      items,
      total
    }
  }
}
