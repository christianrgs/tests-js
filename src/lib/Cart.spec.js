import Cart from './Cart'

const sku = { id: '123', price: 35388 }
const sku2 = { id: '321', price: 41872 }

describe('Cart', () => {
  let cart

  beforeEach(() => {
    cart = new Cart()
  })

  describe('getTotal', () => {
    it('should return 0 when getTotal is executed in a newly created instance', () => {
      expect(cart.getTotal()).toBe(0)
    })

    it('should add item and receive the total amount updated in getTotal', () => {
      cart.add({ sku, quantity: 2 })

      expect(cart.getTotal()).toBe(70776)
    })

    it('should sum the quantity when the product added is already in the cart', () => {
      cart.add({ sku, quantity: 2 })

      expect(cart.getTotal()).toBe(70776)

      cart.add({ sku, quantity: 1 })

      expect(cart.getTotal()).toBe(106164)
    })

    it('should remove item and receive the total amount updated in getTotal', () => {
      cart.add({ sku, quantity: 2 })
      cart.add({ sku: sku2, quantity: 1 })

      expect(cart.getTotal()).toBe(112648)

      cart.remove(sku.id)

      expect(cart.getTotal()).toBe(41872)
    })
  })

  describe('getSummary', () => {
    it('should return an object with the list of items and the total', () => {
      cart.add({ sku, quantity: 2 })
      cart.add({ sku: sku2, quantity: 1 })

      expect(cart.getSummary()).toMatchSnapshot()
      expect(cart.getTotal()).toBeGreaterThan(0)
    })
  })

  describe('checkout', () => {
    it('should return an object with the list of items and the total', () => {
      cart.add({ sku, quantity: 2 })
      cart.add({ sku: sku2, quantity: 1 })

      expect(cart.checkout()).toMatchSnapshot()
    })

    it('should reset the cart when checkout method is called', () => {
      cart.add({ sku, quantity: 3 })

      cart.checkout()

      expect(cart.getTotal()).toBe(0)
    })
  })
})
