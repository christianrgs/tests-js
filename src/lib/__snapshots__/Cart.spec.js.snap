// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart checkout should return an object with the list of items and the total 1`] = `
Object {
  "items": Array [
    Object {
      "quantity": 2,
      "sku": Object {
        "id": "123",
        "price": 35388,
      },
    },
    Object {
      "quantity": 1,
      "sku": Object {
        "id": "321",
        "price": 41872,
      },
    },
  ],
  "total": 112648,
}
`;

exports[`Cart getSummary should return an object with the list of items and the total 1`] = `
Object {
  "items": Array [
    Object {
      "quantity": 2,
      "sku": Object {
        "id": "123",
        "price": 35388,
      },
    },
    Object {
      "quantity": 1,
      "sku": Object {
        "id": "321",
        "price": 41872,
      },
    },
  ],
  "total": 112648,
}
`;
